### PR TITLE
feat: Add IUserSetter and IUserContext

### PR DESCRIPTION
**Pull request by zhenlei520**

### CODE REVIEW

File: src/Identity/Masa.BuildingBlocks.Identity/IUserContext.cs [hunk 1]
```diff
@@ -0,0 +1,9 @@
+﻿// Copyright (c) MASA Stack All rights reserved.
+// Licensed under the MIT License. See LICENSE.txt in the project root for license information.
+
+namespace Masa.BuildingBlocks.Identity;
+
+public interface IUserContext<out TUserId> where TUserId : IComparable
+{
+    TUserId? GetUserId();
+}
```
Comment (doddgu) [inline src/Identity/Masa.BuildingBlocks.Identity/IUserContext.cs:9]: How to get user and user name?

File: src/Identity/Masa.BuildingBlocks.Identity/IUserSetter.cs [hunk 2]
```diff
@@ -0,0 +1,9 @@
+﻿// Copyright (c) MASA Stack All rights reserved.
+// Licensed under the MIT License. See LICENSE.txt in the project root for license information.
+
+namespace Masa.BuildingBlocks.Identity;
+
+public interface IUserSetter<in TUserId> where TUserId : IComparable
+{
+    void SetUserId(TUserId id);
```
Comment (doddgu) [inline src/Identity/Masa.BuildingBlocks.Identity/IUserSetter.cs:8]: How to set user?

**Review by doddgu — CHANGES_REQUESTED** [review]

File: src/Identity/Masa.BuildingBlocks.IdentityModel/DefaultTypeConvertProvider.cs [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+// Licensed under the MIT License. See LICENSE.txt in the project root for license information.
+
+namespace Masa.BuildingBlocks.IdentityModel;
+
+public class DefaultTypeConvertProvider : ITypeConvertProvider
+{
+    public T ConvertTo<T>(string value)
+    {
+        if (typeof(T) == typeof(Guid))
+            return (T)(object)Guid.Parse(value);
```
Comment (doddgu) [inline src/Identity/Masa.BuildingBlocks.IdentityModel/DefaultTypeConvertProvider.cs:11]: `(T)TypeDescriptor.GetConverter(typeof(T)).ConvertFromInvariantString(text);` ?

File: src/Identity/Masa.BuildingBlocks.IdentityModel/IUserContext.cs [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+    string? UserName { get; }
+
+    string? TenantId { get; }
+
+    string? Environment { get; }
+
+    TUserId? GetUserId<TUserId>();
+
+    TTenantId? GetTenantId<TTenantId>();
```
Comment (doddgu) [inline src/Identity/Masa.BuildingBlocks.IdentityModel/IUserContext.cs:22]: Tenant? From ITenant? Like DDD entity use multi UserContext?

**Review by doddgu — CHANGES_REQUESTED** [review]

File: src/Identity/Masa.BuildingBlocks.IdentityModel/IUserContext.cs (hunk 4 above)
Comment (zhenlei520) [reply src/Identity/Masa.BuildingBlocks.IdentityModel/IUserContext.cs:22]: This is not used for multiple DbContexts, but to support different types of TenantId

File: src/Identity/Masa.BuildingBlocks.IdentityModel/DefaultTypeConvertProvider.cs (hunk 3 above)
Comment (zhenlei520) [reply src/Identity/Masa.BuildingBlocks.IdentityModel/DefaultTypeConvertProvider.cs]: processed

File: src/Identity/Masa.BuildingBlocks.IdentityModel/Entities/Isolation/IdentityIsolationUser.cs [hunk 5]
```diff
@@ -3,13 +3,7 @@
 
 namespace Masa.BuildingBlocks.IdentityModel;
 
-public class IdentityUser
+public class IdentityIsolationUser : IdentityMultiTenantUser, IIdentityIsolationUser
```
Comment (doddgu) [inline src/Identity/Masa.BuildingBlocks.IdentityModel/Entities/Isolation/IdentityIsolationUser.cs:6]: `IsolatedIdentityUser`

File: src/Identity/Masa.BuildingBlocks.IdentityModel/Entities/Isolation/IIdentityMultiEnvironmentUser.cs [hunk 6]
```diff
@@ -0,0 +1,9 @@
+﻿// Copyright (c) MASA Stack All rights reserved.
+// Licensed under the MIT License. See LICENSE.txt in the project root for license information.
+
+namespace Masa.BuildingBlocks.IdentityModel;
+
+public interface IIdentityMultiEnvironmentUser : IIdentityUser
```
Comment (doddgu) [inline src/Identity/Masa.BuildingBlocks.IdentityModel/Entities/Isolation/IIdentityMultiEnvironmentUser.cs:6]: `IMultiEnvironmentIdentityUser`

File: src/Identity/Masa.BuildingBlocks.IdentityModel/Entities/Isolation/IIdentityIsolationUser.cs [hunk 7]
```diff
@@ -0,0 +1,8 @@
+﻿// Copyright (c) MASA Stack All rights reserved.
+// Licensed under the MIT License. See LICENSE.txt in the project root for license information.
+
+namespace Masa.BuildingBlocks.IdentityModel;
+
+public interface IIdentityIsolationUser : IIdentityMultiTenantUser, IIdentityMultiEnvironmentUser
```
Comment (doddgu) [inline src/Identity/Masa.BuildingBlocks.IdentityModel/Entities/Isolation/IIdentityIsolationUser.cs:6]: `IsolatedIdentityUser`

File: src/Identity/Masa.BuildingBlocks.IdentityModel/Enum/IdentityType.cs [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+// Licensed under the MIT License. See LICENSE.txt in the project root for license information.
+
+namespace Masa.BuildingBlocks.IdentityModel;
+
+public enum IdentityType
+{
+    /// <summary>
+    /// Only use user information
+    /// </summary>
+    Simple = 1,
```
Comment (doddgu) [inline src/Identity/Masa.BuildingBlocks.IdentityModel/Enum/IdentityType.cs:11]: Basic

File: src/Identity/Masa.BuildingBlocks.IdentityModel/IIsolationUserContext.cs [hunk 9]
```diff
@@ -0,0 +1,10 @@
+﻿// Copyright (c) MASA Stack All rights reserved.
+// Licensed under the MIT License. See LICENSE.txt in the project root for license information.
+
+namespace Masa.BuildingBlocks.IdentityModel;
+
+public interface IIsolationUserContext : IMultiTenantUserContext,
```
Comment (doddgu) [inline src/Identity/Masa.BuildingBlocks.IdentityModel/IIsolationUserContext.cs:6]: Isolated is adj and Isolation is n.

**Review by doddgu — CHANGES_REQUESTED** [review]

File: src/Identity/Masa.BuildingBlocks.IdentityModel/IIsolationUserContext.cs (hunk 9 above)
Comment (zhenlei520) [reply src/Identity/Masa.BuildingBlocks.IdentityModel/IIsolationUserContext.cs]: processed

File: src/Identity/Masa.BuildingBlocks.IdentityModel/Enum/IdentityType.cs [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+
+    /// <summary>
+    /// User Information + Multi-tenant
+    /// </summary>
+    MultiTenant,
```
[the remark this answers is not in the document]
Comment (zhenlei520) [reply src/Identity/Masa.BuildingBlocks.IdentityModel/Enum/IdentityType.cs]: processed

File: src/Identity/Masa.BuildingBlocks.IdentityModel/Enum/IdentityType.cs (hunk 8 above)
Comment (zhenlei520) [reply src/Identity/Masa.BuildingBlocks.IdentityModel/Enum/IdentityType.cs]: processed

File: src/Identity/Masa.BuildingBlocks.IdentityModel/Entities/Isolation/IIdentityMultiTenantUser.cs [hunk 11]
```diff
@@ -0,0 +1,9 @@
+﻿// Copyright (c) MASA Stack All rights reserved.
+// Licensed under the MIT License. See LICENSE.txt in the project root for license information.
+
+namespace Masa.BuildingBlocks.IdentityModel;
+
+public interface IIdentityMultiTenantUser : IIdentityUser
```
[the remark this answers is not in the document]
Comment (zhenlei520) [reply src/Identity/Masa.BuildingBlocks.IdentityModel/Entities/Isolation/IIdentityMultiTenantUser.cs]: processed

File: src/Identity/Masa.BuildingBlocks.IdentityModel/Entities/Isolation/IIdentityIsolationUser.cs (hunk 7 above)
Comment (zhenlei520) [reply src/Identity/Masa.BuildingBlocks.IdentityModel/Entities/Isolation/IIdentityIsolationUser.cs]: processed

File: src/Identity/Masa.BuildingBlocks.IdentityModel/Entities/Isolation/IIdentityMultiEnvironmentUser.cs (hunk 6 above)
Comment (zhenlei520) [reply src/Identity/Masa.BuildingBlocks.IdentityModel/Entities/Isolation/IIdentityMultiEnvironmentUser.cs]: processed

File: src/Identity/Masa.BuildingBlocks.IdentityModel/Entities/Isolation/IdentityIsolationUser.cs (hunk 5 above)
Comment (zhenlei520) [reply src/Identity/Masa.BuildingBlocks.IdentityModel/Entities/Isolation/IdentityIsolationUser.cs]: processed

**Review by doddgu — APPROVED** [review]
LGTM

**Pull request merged**